noir-projects/aztec-nr/aztec/src/encrypted_logs/log_assembly_strategies/default_aes128/note.nr

```diff
@@ -0,0 +1,557 @@
+    assert(ciphertext_bytes.len() == 16 * (1 + ((N * 32) + 64) / 16));
+
+    // *****************************************************************************
+    // Compute the header ciphertext
+    // *****************************************************************************
+
+    let contract_address = context.this_address();
+    let contract_address_bytes = contract_address.to_field().to_be_bytes::<32>();
+
+    let mut header_plaintext: [u8; 32 + 2] = [0; 32 + 2];
+    for i in 0..32 {
+        header_plaintext[i] = contract_address_bytes[i];
+    }
+    let offset = 32;
+    let ciphertext_bytes_length = ciphertext_bytes.len();
+    header_plaintext[offset] = (ciphertext_bytes_length >> 8) as u8;
+    header_plaintext[offset + 1] = ciphertext_bytes_length as u8;
+
+    // TODO: this is insecure and wasteful:
+    // "Insecure", because the esk shouldn't be used twice (once for the header,
+    // and again for the proper ciphertext) (at least, I never got the
+    // "go ahead" that this would be safe, unfortunately).
+    // "Wasteful", because the exact same computation is happening further down.
+    // I'm leaving that 2nd computation where it is, because this 1st computation
+    // will be imminently deleted, when the header logic is deleted.
+    let (sym_key, iv) = derive_aes_symmetric_key_and_iv_from_ecdh_shared_secret_using_sha256(
+        ciphertext_shared_secret,
+    );
+
+    // Note: the aes128_encrypt builtin fn automatically appends bytes to the
+    // input, according to pkcs#7; hence why the output `header_ciphertext_bytes` is 16
+    // bytes larger than the input in this case.
+    let header_ciphertext_bytes = aes128_encrypt(header_plaintext, iv, sym_key);
+    // I recall that converting a slice to an array incurs constraints, so I'll check the length this way instead:
+    assert(header_ciphertext_bytes.len() == HEADER_CIPHERTEXT_SIZE_IN_BYTES);
+
+    // *****************************************************************************
+    // Prepend / append more bytes of data to the ciphertext, before converting back
+    // to fields.
+    // *****************************************************************************
+
+    let mut log_bytes_padding_to_mult_31 =
+        get_arr_of_size__log_bytes_padding__from_PT::<(N * 32) + 64>();
+    /// Safety: this randomness won't be constrained to be random. It's in the
+    /// interest of the executor of this fn to encrypt with random bytes.
+    log_bytes_padding_to_mult_31 = unsafe { get_random_bytes() };
+
+    let mut log_bytes = get_arr_of_size__log_bytes__from_PT::<(N * 32) + 64>();
+
+    assert(
+        log_bytes.len() % 31 == 0,
+        "Unexpected error: log_bytes.len() should be divisible by 31, by construction.",
+    );
+
+    log_bytes[0] = eph_pk_sign_byte;
+    let mut offset = 1;
+    for i in 0..header_ciphertext_bytes.len() {
+        log_bytes[offset + i] = header_ciphertext_bytes[i];
+    }
+    offset += header_ciphertext_bytes.len();
```

Comment:
We need a concat or insert fn yesterday, this is a recipe for disaster

Reply:
https://github.com/noir-lang/noir/pull/7199/files :)